software/contrib/hamlet.py

```diff
@@ -0,0 +1,359 @@
+                self.gate_1.off()
+                
+            if track_2_sparsity > self.sparsity:
+                self.cv_2.voltage(track_2_cv)
+                self.gate_2.on()
+            else:
+                self.gate_2.off()
+
+            # Reset clock step at 128 to avoid a HUGE integer if running for a long time
+            # over a really long period of time this would look like a memory leak
+            if self.clock_step < 128:
+                self.clock_step +=1
+            else:
+                self.clock_step = 0
+    
+            # Reset step number at step_length -1 as pattern arrays are zero-based
+            if self.drum_step < self.step_length - 1:
+                self.drum_step += 1
+            else:
+                self.drum_step = 0
+
+            if self.track_step < (self.step_length * self.track_multiplier) - 1:
+                self.track_step += 1
+            else:
+                self.track_step = 0
+
+                
+        @din.handler_falling
+        def clockTriggerEnd():
+            self.bd.off()
+            self.hh.off()
+            self.gate_1.off()
+            self.gate_2.off()
+
+    def generateNewRandomCVPattern(self):
+        self.step_length = len(self.BD[self.pattern])
+        # CV patterns are up to 4 times the length of the drum pattern
+        patt = (self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9))
+        self.track_1.append(patt)
+        patt = (self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9) +
+                  self.generateRandomPattern(self.step_length, 0, 9))
+        self.track_2.append(patt)
+
+    def updatePattern(self):
+        # If mode 2 and there is CV on the analogue input use it, if
+        # not use the knob position
+        val = 100 * ain.percent()
+        if self.analogInputMode == 2 and val > self.minAnalogInputVoltage:
+            self.pattern = int((len(self.BD) / 100) * val)
+        else:
+            self.pattern = k2.read_position(len(self.BD))
+        
+        self.step_length = len(self.BD[self.pattern])
+
+    def updateCvPattern(self):
+        # If analogue input mode 3, get the CV pattern from CV input
+        if self.analogInputMode != 3:
+            return
+        else:
+            # Get the analogue input voltage as a percentage
+            CvpVal = 100 * ain.percent()
+            
+            # Is there a voltage on the analogue input and are we configured to use it?
+            if CvpVal > 0.4:
+                # Convert percentage value to a representative index of the pattern array
+                self.CvPattern = int((len(self.track_1) / 100) * CvpVal)
```

Comment:
Optional: the `else` is unnecessary since the `if` condition exits the method if true, so this allows you to un-indent this block of code to help readability.
```suggestion
       
        # Get the analogue input voltage as a percentage
        CvpVal = 100 * ain.percent()
        
        # Is there a voltage on the analogue input and are we configured to use it?
        if CvpVal > 0.4:
            # Convert percentage value to a representative index of the pattern array
            self.CvPattern = int((len(self.track_1) / 100) * CvpVal)
```